src/main/java/sample/ClientCredentialsAuthenticationFilter.java

```diff
@@ -0,0 +1,123 @@
+ * limitations under the License.
+ */
+
+package sample;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.Base64;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * A filter to perform client authentication for the Token Endpoint, specifically for the Client Credentials Grant.
+ *
+ * See <a href="https://tools.ietf.org/html/rfc6749#section-2.3.1">RFC-6749 2.3.1</a>.
+ */
+public class ClientCredentialsAuthenticationFilter implements Filter {
+
+    private Charset credentialsCharset = UTF_8;
+
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+
+        HttpServletRequest req = (HttpServletRequest) request;
+        if (req.getRequestURI().equals("/oauth/token")) {
+            String grantType = req.getParameter("grant_type");
+            if ("client_credentials".equals(grantType)) {
+                String[] credentials = extractCredentials(req);
+                String clientId = credentials[0];
+                String clientSecret = credentials[1];
+                if (!clientId.equals("myclientid") || !clientSecret.equals("myclientsecret")) {
+                    throw new BadCredentialsException("Invalid client credentials");
+                }
+                req.getSession().setAttribute("authentication", true);
+            }
+        }
+
+        chain.doFilter(request, response);
+    }
+
+    private String[] extractCredentials(HttpServletRequest request) {
+        String[] credentials = extractBasicAuthenticationCredentials(request);
+        if (credentials == null) {
+            credentials = extractParameterAuthenticationCredentials(request);
+        }
+        if (credentials == null) {
+            throw new InvalidRequestException("Invalid token request");
+        }
+        return credentials;
+    }
+
+    private String[] extractBasicAuthenticationCredentials(HttpServletRequest request) {
+        String header = request.getHeader("Authorization");
+        if (header != null && header.toLowerCase().startsWith("basic ")) {
+            return extractAndDecodeHeader(header, request);
+        }
+        return null;
+    }
+
+    private String[] extractParameterAuthenticationCredentials(HttpServletRequest request) {
+        String clientId = request.getParameter("client_id");
+        if (clientId != null) {
+            String clientSecret = request.getParameter("client_secret");
+            if (clientSecret == null) {
+                throw new InvalidRequestException("No client_secret parameter provided");
+            }
+            return new String[] { clientId, clientSecret };
+        }
+        return null;
+    }
+
+    // Taken from BasicAuthenticationFilter (spring-security-web)
+    private String[] extractAndDecodeHeader(String header, HttpServletRequest request) {
```

Comment:
why give the header name as parameter? At least I do only know one header ("Authorization") that may contain the content that is necessary for this method to operate successfully. If I am wrong please correct me.